### PR TITLE
Adding DevOps with GitHub v1

### PR DESCRIPTION
**Pull request by kevinmgates**
Adding DevOps with GitHub. Will update in the future with coach guides.

### CODE REVIEW

File: 031-DevOpsWithGitHub/readme.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+1. How to use GitHub Actions for CI & CD
+1. Monitoring apps with Application Insights
+
+## Challenges
+ - [Challenge 0](./Student/challenge00.md) - Setup and Introduction
+ - [Challenge 1](./Student/challenge01.md) - Track your work with GitHub Project Boards
+ - [Challenge 2](./Student/challenge02.md) - Centralize your code with GitHub Repos
+ - [Challenge 3](./Student/challenge03.md) - Infrastructure as Code
+ - [Challenge 4](./Student/challenge04.md) - Continuous Integration
+ - [Challenge 5](./Student/challenge04.md) - Build and push Docker image to container registry
```
Comment (jrzyshr) [inline 031-DevOpsWithGitHub/readme.md:21]: Challenges 5 through 9 link to challenge N-1 ;)  Please fix!

File: 031-DevOpsWithGitHub/Student/Template.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
+# What The Hack - Challenge Template
```
Comment (jrzyshr) [inline 031-DevOpsWithGitHub/Student/Template.md:1]: Please remove this template file!

File: 031-DevOpsWithGitHub/Student/challenge02.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+### Challenge
+
+Now that we have a basic understanding of version control and Git, lets get some code checked into source control. DevOps best practices can apply to any programming language, so for today we have provided you a simple .NET Core web application to use.
+
+1. Begin by cloning the GitHub repository you created in the [first challenge](challenge01.md) to your local computer ([hint](https://help.github.com/en/articles/cloning-a-repository)).
+
+2. Next, download the code for the application [here](https://github.com/OCPDemos/WhatTheHack/blob/027-DevOps-with-GitHub/027-DevOpsWithGitHub/Student/Code/Code.zip). Unzip the download and copy the contents into the root of the repository you just cloned.
```
Comment (jrzyshr) [inline 031-DevOpsWithGitHub/Student/challenge02.md:17]: We discussed this on our review call.  Zip files should not be stored in the GH repo. 

We previously would have recommended using DownGit, but it's been glitchy, and occasionally blocked by Chrome as malware!

New recommendation is to ensure the lead coach zips up the contents of the /Student/Resources folder and post it to the Files tab of the hack event's team.   Here in the challenge text, it should say something like "Obtain the code from the Resource files provided by your coach"

**Review by jrzyshr — CHANGES_REQUESTED** [review]
Thanks for this incoming hack!  Looks great.  Just a few minor changes I'm requesting in the comments inline in the files.

Also, the /Student/Code folder should be renamed to /Student/Resources

File: 031-DevOpsWithGitHub/readme.md (hunk 1 above)
Comment (kevinmgates) [reply 031-DevOpsWithGitHub/readme.md]: Thanks...fixed!

File: 031-DevOpsWithGitHub/Student/Template.md (hunk 2 above)
Comment (kevinmgates) [reply 031-DevOpsWithGitHub/Student/Template.md]: Removed!

File: 031-DevOpsWithGitHub/Student/challenge02.md (hunk 3 above)
Comment (kevinmgates) [reply 031-DevOpsWithGitHub/Student/challenge02.md]: I removed the .zip file itself but forgot about that reference. Fixed!

File: 031-DevOpsWithGitHub/Student/challenge08.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+2. Create a dashboard in the Azure Portal to provide a summary of the status of our site. ([hint](https://docs.microsoft.com/en-us/azure/azure-monitor/app/overview-dashboard#application-dashboard))
+
+3. Implement an outside in availability test for the homepage of your site ([hint](https://docs.microsoft.com/en-us/azure/azure-monitor/app/monitor-web-app-availability))
+
+### Success Criteria
+
+1. You should understand the importance of monitoring and some of the basic features offered by Application Insights.
+
+    > NOTE: We are just scratching the surface of what is offered in Azure Monitoring, if you are interested in learning more there is a full What the Hack focused on [Azure Monitoring](https://github.com/microsoft/WhatTheHack/tree/master/007-AzureMonitoring).
```
Comment (gfilicetti) [inline 031-DevOpsWithGitHub/Student/challenge08.md:28]: Please don't use the '>' quote markdown syntax here, just make these sub-bullets (and bold up the NOTE, :-) ).

File: 031-DevOpsWithGitHub/Student/challenge08.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+### Advanced Challenges (optional)
+
+1. If you integrated Azure DevOps Boards with GitHub, link your Application Insights instance with your Azure Boards instance ([hint](https://azure.microsoft.com/en-us/blog/application-insights-work-item-integration-with-visual-studio-team-services/))
+
+2. Using the failures feature of Application Insights find an exception that happened on your site. Using the link you created in the last step, open a work item to resolve that exception. 
+
+    > NOTE: If your site doesn’t have any exceptions, you can create one easily by trying to go to a page that doesn’t exist.
```
Comment (gfilicetti) [inline 031-DevOpsWithGitHub/Student/challenge08.md:36]: Same as above re: quote

File: 031-DevOpsWithGitHub/Student/challenge08.md [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+    > NOTE: It takes a min or two after an event happens for it to make its way to Application Insights and for it to be indexed so you can see it in the portal.
```
Comment (gfilicetti) [inline 031-DevOpsWithGitHub/Student/challenge08.md:38]: Same as above re: quote

File: 031-DevOpsWithGitHub/Student/challenge09.md [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+### Challenge
+
+In this challenge, you will improve the security of your repository using some of GitHub's built-in tools. 
+
+1. Find the repository's Security policy. If there is an existing policy, make an edit and merge your change back into the main branch. Otherwise, go ahead and create a policy using the template provided. GitHub Security policies are Markdown documents that indicate the preferred way to report security vulnerabilities for the repository. 
+
+2. Enable Dependabot alerts for the repository. Dependabot is an automated tool that creates a pull request when any dependencies in the code base has a known vulnerability. 
+
+3. (NOTE: Skip this step while code scanning is in Beta.) Finally, set up and run a Code scanning workflow for the repository using GitHub's 'CodeQL Analysis.' This workflow can run either on each pull request or on a schedule, and it checks your code for common vulnerabilities or errors.
```
Comment (gfilicetti) [inline 031-DevOpsWithGitHub/Student/challenge09.md:23]: Bold.

**Review by gfilicetti — CHANGES_REQUESTED** [review]
Great stuff overall, just a bunch of tiny nitpicks cause I'm silly like that.

File: 031-DevOpsWithGitHub/Coach/readme.md [hunk 8]
```diff
@@ -0,0 +1,3 @@
+# What the Hack: DevOps with GitHub
+
+## The Proctor's Guide and Lecture slide decks
```
[the remark this answers is not in the document]
Comment (jrzyshr) [reply 031-DevOpsWithGitHub/Coach/readme.md:3]: I'm okay with this.  A coach guide can always be added afterwards as a post-mortem after the first event.

[the remark this answers is not in the document]
Comment (gfilicetti) [reply 031-DevOpsWithGitHub/Coach/readme.md:3]: I'm ok if we go live without a coach's guide as well, but in that case please delete this readme and the Coach folder completely.

**Review by gfilicetti — CHANGES_REQUESTED** [review]

File: 031-DevOpsWithGitHub/readme.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+ - [Challenge 8](./Student/challenge07.md) - Monitoring: Application Insights
+ - [Challenge 9](./Student/challenge08.md) - Security
+
+
+## Prerequisites
+- Your own Azure subscription with **owner** access. See considerations below for additional guidance.
+- [Visual Studio Code](https://code.visualstudio.com)
+- [Git SCM](https://git-scm.com/download)
+
+## Repository Contents (Optional)
```
[the remark this answers is not in the document]
Comment (kevinmgates) [reply 031-DevOpsWithGitHub/readme.md:33]: Done!

File: 031-DevOpsWithGitHub/Coach/readme.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (kevinmgates) [reply 031-DevOpsWithGitHub/Coach/readme.md:3]: Will do

File: 031-DevOpsWithGitHub/Student/challenge03.md [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    - When you have successfully created the `staging` environment, you can delete it as it will not be used in the upcoming challenges. 
+
+2. In this challenge, we setup three separate workflow files to handle `dev`, `test` and `prod`, however, this could be done with a single workflow with the target environment stored as a secret instead of hardcoding it in the workflow file.
+    - Create a GitHub secret (called `targetEnv`) and set the *value* to `staging2` ([hint](https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets))
+    - In your workflow, read the GitHub secret ([hint](https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets#using-encrypted-secrets-in-a-workflow))
+and pass the value as a overriding template parameter (as described in Advanced Challenge #1)
+    - When you have successfully created the `staging2` environment, you can delete it as it will not be used in the upcoming challenges.
+
+
+NOTE: If you are interested in learning more about Infrastructure as Code, there are [multiple](https://github.com/microsoft/WhatTheHack) What the Hacks that cover it in greater depth.
```
[the remark this answers is not in the document]
Comment (kevinmgates) [reply 031-DevOpsWithGitHub/Student/challenge03.md:76]: Makes sense. Done!

File: 031-DevOpsWithGitHub/Student/challenge05.md [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    - `repositoryName` - The repository to target in the registry. Set this to "`wth/dotnetcoreapp`".
+    - `dockerFolderPath` - The path to the folder that contains the Dockerfile - a critical parameter. You will need to point to the folder: `Application/aspnet-core-dotnet-core`.
+    - `tag` - This needs to be a unique value each time, as this is used to version the images in the repository. GitHub makes [environment variables](https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions#github-context) available that helps with this. Set `tag` to the [github.run_number](https://www.bing.com/search?q=%24%7B%7Bgithub.run_number%7D%7D&form=QBLH&sp=-1&pq=%24%7B%7Bgithub.run_number%7D%7D&sc=0-22&qs=n&sk=&cvid=D84DA66323DC4E14BD794F90FCFD90D3) environment variable.
+
+2. Go to the Azure Portal and get the (1) username and (2) password and (3) login server to your ACR instance and save as GitHub secrets (ACR_USERNAME, ACR_PASSWORD, ACR_LOGIN_SERVER).
```
[the remark this answers is not in the document]
Comment (kevinmgates) [reply 031-DevOpsWithGitHub/Student/challenge05.md:30]: Done!

File: 031-DevOpsWithGitHub/Student/challenge06.md [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+In DevOps after we automate our build process, we want to automate our release process, we do this with a technique called Continuous Delivery (CD). Please take a moment to review this brief article talking about why this is important. 
+
+- [What is Continuous Delivery?](https://docs.microsoft.com/en-us/azure/devops/learn/what-is-continuous-delivery)
+
+### Challenge
+
+In this challenge, we will use GitHub Actions to deploy our container image to the dev environment. 
+
+OPTIONAL: Use your code editor (VS Code) to update your workflow file locally on your machine. Remember to commit and push any changes.
```
[the remark this answers is not in the document]
Comment (kevinmgates) [reply 031-DevOpsWithGitHub/Student/challenge06.md:17]: Done

File: 031-DevOpsWithGitHub/Student/challenge07.md [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+- [About required reviews for pull requests](https://docs.github.com/en/free-pro-team@latest/github/administering-a-repository/about-required-reviews-for-pull-requests)
+
+### Tips
+
+- If your GitHub account was created on the 'Free' tier, then in order to create a Branch Protection rule your repository must be public. To change a repository from private to public, visit the 'Settings' tab, and scroll to the bottom where you have the option to 'Change visibility.'
+- If using the git command line interface, you can find a number of sample git commands that are useful for branching [here](https://gist.github.com/JamesMGreene/cdd0ac49f90c987e45ac). (Make sure to focus on the 'git' commands, rather than 'gitflow'.)
+- If using the git command line interface, try adding '--help' after a command to get helpful information about arguments and usage.
+
+
+### Advanced Challenges
```
[the remark this answers is not in the document]
Comment (kevinmgates) [reply 031-DevOpsWithGitHub/Student/challenge07.md:53]: Thanks for catching that. Done.

File: 031-DevOpsWithGitHub/Student/challenge07.md [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+
+In this challenge, we focused on creating a feature branch directly off of the main branch. Some organizations, however, prefer to do phased deployments. Instead of merging feature branches directly back into production, this alternate strategy involves having a main production branch and a development branch which runs parallel to the main branch. Feature and bug fix branches are created from and merged into the development branch. When you want to release new features to production, create a pull request to merge changes from development into the main branch. 
+
+If you would like to explore this flow, try to set up your repository for these 'phased deployments.' Begin by creating a development branch off of your main branch. On the development branch, repeat the flow from above. When you are ready to release, create and complete a pull request merging the development branch into the main branch. 
+
+IMPORTANT: Do not delete the development branch after completing the deployment. You will want to use this same branch to repeat the process for your next deployment.
```
[the remark this answers is not in the document]
Comment (kevinmgates) [reply 031-DevOpsWithGitHub/Student/challenge07.md:59]: Fixed!

File: 031-DevOpsWithGitHub/Student/challenge08.md (hunk 4 above)
Comment (kevinmgates) [reply 031-DevOpsWithGitHub/Student/challenge08.md:28]: Makes sense!

**Review by gfilicetti — APPROVED** [review]
Great work @kevinmgates !! Thanks for putting up with the nit picking

**Review by jrzyshr — APPROVED** [review]
Thanks for the contributions Kevin & Will & Julia!!!!

**Pull request merged**